metrics: Add metrics to report count of current endpoints tagged by states

manalibhutiyani opened the pull request:
This change adds a new prometheus gauge metrics to expose the
count of current endpoints labelled by various endpoint states.

Fixes: #3863
Signed-Off-By: Manali Bhutiyani <manali@covalent.io>


```release-note
Metrics to report count of current endpoints tagged by endpoint states
```

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/4376)
<!-- Reviewable:end -->

Commit:
metrics: Add metrics to report count of current endpoints tagged by states.

This commit adds a new prometheus gauge metrics to expose the
count of current endpoints labelled by various endpoint states.
This commit also merges all endpoint states to be set correctly using
SetStateLocked()

Fixes: #3863
Signed-Off-By: Manali Bhutiyani <manali@covalent.io>

rlenglet commented on pkg/metrics/metrics.go:
```diff
@@ -78,6 +78,16 @@ var (
+			Name:      "endpoint_state",
+			Help:      "Count of all endpoints, tagged by different endpoint states",
+		},
+		[]string{"endpoint state"},
```
Per the Prometheus naming conventions, it seems label names shouldn't contain spaces:
https://prometheus.io/docs/practices/naming/

So I'd suggest either `endpoint_state` or just `state`.

manalibhutiyani replied:
ah.. I guess labels are allowed to have spaces but not label names. Let me change that

manalibhutiyani replied:
done. Added special case for cilium health as well. please check

rlenglet reviewed (dismissed):
You're missing counting the initial states of endpoints.
You'll need to modify `createEndpoint` in `daemon/endpoint.go` to increment the count for the endpoint's initial state.

Likewise, you'll need to decrement when an endpoint is deleted.

Commit:
docs: Add the endpoint state count metrics to docs

Fixes: #3863
Signed-Off-By: Manali Bhutiyani <manali@covalent.io>

raybejjani reviewed (dismissed):
Thinking about it a little, I realised that it doesn't help if we have a separate metric that has these labels. They would both be collected by prometheus (and so the cardinality of the labels would matter). Since the state is one of the more important aspects of an endpoint I would say it's ok to put the labels on the `EndpointCount` gauge. I'm not sure how to do that with a `GaugeFunc`, though. It might not be possible without splitting it as we do with the `eventTS` metrics.

Catching the state change is pretty slick, though, so perhaps we could replace the GaugeFunc with this method. My only concern is that we might not call `SetState` when we create an endpoint so we would undercount new endpoints that are still in `init` or `restoring` (or whatever is equivalent). This would be straightforward to fix, I think, since we could add a dummy transition to record this (this would make the endpoint log more correct too!).
Another thing would be whether we are ok with these labels not including `not-ready`. We synthesise this state in `GetModel` (https://github.com/cilium/cilium/blob/master/pkg/endpoint/endpoint.go#L688-L691) and `GetHealthModel` (https://github.com/cilium/cilium/blob/master/pkg/endpoint/endpoint.go#L770-L772). Unfortunately, we present this as a real state to users in the docs so it will be confusing if they cannot filter for a `not-ready` label. Worse, this won't be apparent since most of the time endpoints are not `not-ready` so debugging something like that would be frustrating.

tgraf commented on daemon/endpoint.go:
```diff
@@ -373,6 +377,13 @@ func (d *Daemon) deleteEndpoint(ep *endpoint.Endpoint) int {
+
+	// Since the endpoint is already removed from endpointmanager queue,
+	// we do not need to acquire any lock on it.
+	if len(errors) == 0 {
```
Move this to where we call `endpointmanager.Remove()` in `deleteEndpointQuiet()` and do not make it conditional on no errors. The endpoint is removed regardless of whether there have been errors. Otherwise you are leaking endpoint counts.

raybejjani replied:
This might be even more complicated. `deleteEndpointQuiet` can return early if the endpoint state is disconnecting. I'm not sure if we re-remove it later. In any case, that return happens before the `endpointmanager.Remove` so we would leak disconnecting endpoints.

manalibhutiyani replied:
@tgraf : I am trying to understand how it matters if this is called here or in `endpointmanager.Remove()` . I understand the part of not making it conditional with Len(errors), but at this point we would have already called `deleteEndpointQuiet` and removed the endpoint safely from `endpointmanager`. We can simply update the metrics safely at this point? What's the rationale of moving this code to `endpointmanager.Remove() `in `deleteEndpointQuiet()`

manalibhutiyani replied:
Additionally we do not need any locking at this point, since the endpoint is no longer in `endpointmanager queue`

tgraf replied:
As the code is as-is, there is no difference but what if endpoints are added and deleted via something other than the API in the future? The endpointmanager controls the lifecycle of an endpoint and the insertion and removal is gated via `endpointmanager.Insert()` and `endpointmanager.Remove()`. Bumping metrics there seems more logical to me.

manalibhutiyani replied:
Ah ok..got it! thanks. I will move it there.

manalibhutiyani replied:
done.

Commit:
metrics: Deprecate EndpointCountRegenerating in favor of EndpointStateCount

This commit deprecates the singular metrics `EndpointCountRegenerating` in favor of
a more generic EndpointStateCount representing all the endpoint states, labelled with
necessary states. We are still keeping this metrics around for backward compatibility  and
simply changing the docs and description to warn existing users.

Fixes: #3863
Signed-Off-By: Manali Bhutiyani <manali@covalent.io>

Commit:
docs: Deprecate EndpointCountRegenerating in favor of EndpointStateCount

This commit deprecates the singular metrics `EndpointCountRegenerating` in favor of
a more generic EndpointStateCount representing all the endpoint states, labelled with
necessary states. We are still keeping this metrics around for backward compatibility  and
simply changing the docs and description to warn existing users.

Fixes: #3863
Signed-Off-By: Manali Bhutiyani <manali@covalent.io>

aanm commented on daemon/endpoint.go:
```diff
@@ -177,6 +180,13 @@ func (d *Daemon) createEndpoint(epTemplate *models.EndpointChangeRequest, id str
+	// we need to increment the counters before `SetIdentityLabels`
+	// since `SetIdentityLabels` can trigger an endpoint regenerate
+	// and change the state to "waiting-to-regenerate"
+	metrics.EndpointStateCount.
```
I don't understand the comment. How do you know the endpoint state is `waiting-for-identity`?

manalibhutiyani replied:
Since this is called in `createEndpoint` the comment states that in order to capture the initial state `waiting-for-identity` we need to call it before we set the identity.

manalibhutiyani replied:
I can change the wording to more generic:
```
// In order to capture initial states
// we need to increment the counters before `SetIdentityLabels`
// since `SetIdentityLabels` can trigger an endpoint regenerate
// and change the state to "waiting-to-regenerate"
```

joestringer reviewed (dismissed):
Please also run `promtool check-metrics` against these new metrics as well.

rlenglet commented on cilium-health/launch/endpoint.go:
```diff
@@ -41,6 +41,7 @@ import (
 	"github.com/cilium/cilium/pkg/option"
 	"github.com/cilium/cilium/pkg/pidfile"
 
+	"github.com/cilium/cilium/pkg/metrics"
```
Move this import to the import group above.

manalibhutiyani replied:
sure

rlenglet reviewed (approved):
LGTM. Just one nit.

tgraf commented on daemon/endpoint.go:
```diff
@@ -177,6 +180,13 @@ func (d *Daemon) createEndpoint(epTemplate *models.EndpointChangeRequest, id str
+	// counters before `SetIdentityLabels` since `SetIdentityLabels` can
+	// trigger an endpoint regenerate
+	// and change the state to "waiting-to-regenerate"
+	metrics.EndpointStateCount.
```
I'm not sure this is correct. `PUT /endpoint/{id}` can be called on existing endpoints. It is clearly not symmetric. Why can't we solve this entirely in `SetState` and detect the initial state change `nil -> ...` and the final `... -> disconnected` to do proper accounting?

manalibhutiyani replied:
@tgraf `EndpointStateWaitingForIdentity` identity state needs to be captured in `createEndpoint` I never see this state change in  `SetState` i.e toState is never `EndpointStateWaitingForIdentity` and hence my understanding was that this initial state needs to be captured in `createEndpoint`.  CC ing to confirm this: @rlenglet 

Also, we do not strictly follow `SetState` to set endpoint states.. I can see this in case of `EndpointStateWaitingForIdentity` state as well as `restoring` state.

rlenglet replied:
@tgraf When `PUT /endpoint/{id}` is called on an existing endpoint, that existing endpoint is first deleted:
```
		e, err := endpointmanager.Lookup(params.ID)
		if err != nil {
			// Delete endpoint if any operation that occurs during PUT, fails,
			// so that endpoints in an unhealthy state are not left lying around.
			if e != nil {
				h.d.deleteEndpoint(e)
			}
			return api.Error(PutEndpointIDFailedCode, err)
		}
```

I think nevertheless that calling `SetState` instead of directly modifying `e.state` directly is a good thing to do. There are currently 3 places where `Endpoint.state` is set without calling `SetStateLocked`. Those places need to be modified to call `SetStateLocked` instead.

tgraf replied:
The code you quote does not delete the endpoint when it already exists. It handles the case when the endpoint was removed while we are waiting for the endpoint to be built in sync mode.

manalibhutiyani replied:
@tgraf Done. Please take a look. Also handled special cases for `disconnected` state and initial `nil->waiting-for-identity` state. Audited and `e.state` is set now only via calling `SetState*()`

tgraf commented on pkg/endpoint/endpoint.go:
```diff
@@ -1962,6 +1970,20 @@ func (e *Endpoint) SetStateLocked(toState, reason string) bool {
+
+	// Initial state transitions i.e nil --> waiting-for-identity
+	// need to be handled correctly while updating metrics.
+	if fromState != "" {
```
This only works for new endpoints. If we restore endpoints, the fromState will be "" but there is no previous increment operation to compensate for. The code as-is will result in negative state counters for restore endpoints when Cilium is restarted. I think you have to special case the restore operation as the previous state as recovered in `ParseEndpoint()` is not represented in the in-memory metrics. This also means that you can undo the additions to the valid state transitions above.

manalibhutiyani replied:
You mean I need to add a special case for `StateRestoring`

Would this explicit check make more sense then?

```
       // Initial state transitions i.e nil --> waiting-for-identity
	// need to be handled correctly while updating metrics.
	// Note that if we are transitioning from some state to restoring
	// state, we cannot decrement the old state counters as they will not
	// be accounted for in the metrics.
	if fromState != "" && toState != StateRestoring {
		metrics.EndpointStateCount.
			WithLabelValues(fromState).Dec()
	}
```

manalibhutiyani replied:
I am unclear about the second half of the comment, but the reason I have added StateRestoring in the state transitions above in the switch case is to make the code jump to `OKState:` because we still need to execute this part of the code for ` StateRestoring`
```
// Since StateDisconnected is the final state, after which the
	// endpoint is gone, we should not increment metrics for this state.
	if toState != "" && toState != StateDisconnected {
		metrics.EndpointStateCount.
			WithLabelValues(toState).Inc()
	}
```
 i.e increment the `StateRestoring` part? Let me know if my understanding is correct here.. I hope I have interpreted your review comment correctly.

manalibhutiyani replied:
done.. not sure if it addresses all your concerns, please let me know if I have not understood something.
